File: lib/xycrypt.js
// Code to decrypt pokemon pkx files
// Algorithm by Xfr and Bond697
var Long = require('long')

var PkxSizes = {
    SIZE_PARTY_PKX      : 260,
    SIZE_BOX_PKX        : 232,
    SIZE_PKX_UNCRYPT    : 8,
    SIZE_PKX_CRYPT_LEN  : 232 - 8,
    SIZE_PKX_BLK        : 56,
    SIZE_PKX_PARTY_DATA : 28
}

function Rnd(data) {
    this.mul  = 0x41C64E6D
    this.add  = 0x6073
    this.rmul = 0xEEB9EB65
    this.radd = 0xA3561A1
    this.data = data
}

Rnd.prototype.advance = function() {
    var data_l = new Long(this.data, 0)
    var mul_l = new Long(this.mul, 0)
    this.data = (data_l.multiply(mul_l).low + this.add) & 0xFFFFFFFF
    return (this.data >> 16) & 0xFFFF
}

var orders = [
    [0, 1, 2, 3],
    [0, 1, 3, 2],
    [0, 2, 1, 3],
    [0, 3, 1, 2],
    [0, 2, 3, 1],
    [0, 3, 2, 1],
    [1, 0, 2, 3],
    [1, 0, 3, 2],
    [2, 0, 1, 3],
    [3, 0, 1, 2],
    [2, 0, 3, 1],
    [3, 0, 2, 1],
    [1, 2, 0, 3],
    [1, 3, 0, 2],
    [2, 1, 0, 3],
    [3, 1, 0, 2],
    [2, 3, 0, 1],
    [3, 2, 0, 1],
    [1, 2, 3, 0],
    [1, 3, 2, 0],
    [2, 1, 3, 0],
    [3, 1, 2, 0],
    [2, 3, 1, 0],
    [3, 2, 1, 0]
];


function decrypt(in_pkx) {
    var box_pkx = new Buffer(PkxSizes.SIZE_BOX_PKX)
    var outbox_pkx = new Buffer(PkxSizes.SIZE_BOX_PKX)
    var party_data = new Buffer(PkxSizes.SIZE_PKX_PARTY_DATA)
    var out_pkx = new Buffer(in_pkx.length)

    in_pkx.copy(box_pkx, 0, 0, PkxSizes.SIZE_PARTY_PKX)

    if ((in_pkx.length != PkxSizes.SIZE_PARTY_PKX) && (in_pkx.length != PkxSizes.SIZE_BOX_PKX))
    {
        throw new Error("Pkx is the wrong size.");
    }


    var decryptKey = box_pkx.readUInt32LE(0)

    CryptPkx(box_pkx.slice(8), decryptKey)
    UnshufflePkx(box_pkx, outbox_pkx);
    SetUncryptedDataEqual(box_pkx, outbox_pkx);

    if (!VerifyPkxChecksum(box_pkx.slice(8), box_pkx.readUInt16LE(6), PkxSizes.SIZE_PKX_CRYPT_LEN))
    {
        throw new Error("Checksum failed.")
    }

    if (in_pkx.length == PkxSizes.SIZE_PARTY_PKX)
    {
        decryptKey = box_pkx.readUInt32LE(0)

        in_pkx.copy(party_data, PkxSizes.SIZE_BOX_PKX, 0, PkxSizes.SIZE_PKX_PARTY_DATA)

        CryptPkx(party_data, decryptKey);
    }

    outbox_pkx.copy(out_pkx, 0, 0, PkxSizes.SIZE_BOX_PKX)

    if (in_pkx.length == PkxSizes.SIZE_PARTY_PKX)
    {
        party_data.copy(out_pkx, 0, PkxSizes.SIZE_BOX_PKX, PkxSizes.SIZE_PKX_PARTY_DATA)
    }

    return out_pkx
}

function CryptPkx(pkx, key) {
    var cryptKey = 0, i, len = pkx.length
    var cr = new Rnd(key)
    for (i = 0; i < len; i+=2) {
        cryptKey = cr.advance()
        pkx.writeUInt16LE(pkx.readUInt16LE(i) ^ cryptKey, i)
    }
}

function CalcPkxChecksum(pkx, len) {
    var sum = 0, i
    for (i = 0; i < len; i+=2) {
        sum += pkx.readUInt16LE(i)
    }
    return sum & 0xFFFF
}

function VerifyPkxChecksum(pkx, currentSum, len) {
    var checksum = CalcPkxChecksum(pkx, len)
    return checksum === currentSum
}

function FixPkxChecksum(pkx, len) {
    var sum = CalcPkxCheckSum(pkx.slice(8), len)
    pkx.writeUInt16LE(sum, 6)
}

function GetShuffleType(pid) {
    return ((pid & 0x3E000) >> 0xD) % 24;
}

function UnshufflePkx(pkx_in, pkx_out) {
    var blockOrder = GetShuffleType(pkx_in.readUInt32LE(0)), i
    for (i = 0; i < 4; ++i) {
        var sourceStart = PkxSizes.SIZE_PKX_UNCRYPT + (PkxSizes.SIZE_PKX_BLK * orders[blockOrder][i])
        pkx_in.copy(pkx_out, PkxSizes.SIZE_PKX_UNCRYPT + (PkxSizes.SIZE_PKX_BLK * i),
                sourceStart, sourceStart + PkxSizes.SIZE_PKX_BLK)
    }
}

function SetUncryptedDataEqual(pkx_in, pkx_out) {
    pkx_in.copy(pkx_out, 0, 0, 8)
}

exports.decrypt = decrypt

File: src/xycrypt.cc
#define _CRT_SECURE_NO_WARNINGS

#ifdef WINDOWS
#include <windows.h>
#else
#include <cstring>
#endif
#include <stdio.h>
#include <stdlib.h>
#include "dtypes.h"
#include "xycrypt.h"


enum PkxSizes {

    SIZE_PARTY_PKX      = 260,
    SIZE_BOX_PKX        = 232,
    SIZE_PKX_UNCRYPT    = 8,
    SIZE_PKX_CRYPT_LEN  = SIZE_BOX_PKX - SIZE_PKX_UNCRYPT,
    SIZE_PKX_BLK        = 56,
    SIZE_PKX_PARTY_DATA = 28

};


struct Rnd {

    u32 mul;
    u32 add;
    u32 rmul;
    u32 radd;

};


const char orders[24][4] = {
    0, 1, 2, 3,
    0, 1, 3, 2,
    0, 2, 1, 3,
    0, 3, 1, 2,
    0, 2, 3, 1,
    0, 3, 2, 1,
    1, 0, 2, 3,
    1, 0, 3, 2,
    2, 0, 1, 3,
    3, 0, 1, 2,
    2, 0, 3, 1,
    3, 0, 2, 1,
    1, 2, 0, 3,
    1, 3, 0, 2,
    2, 1, 0, 3,
    3, 1, 0, 2,
    2, 3, 0, 1,
    3, 2, 0, 1,
    1, 2, 3, 0,
    1, 3, 2, 0,
    2, 1, 3, 0,
    3, 1, 2, 0,
    2, 3, 1, 0,
    3, 2, 1, 0,
};


inline void InitRnd(Rnd* r);
inline u16 CryptRNG(u32* data, Rnd* r);
void CryptPkx(u32* pkx, u32* key, u32 len, Rnd* cr);
inline u16 CalcPkxChecksum(u16* pkx, u32 len);
bool VerifyPkxChecksum(u32* pkx, u16 currentSum, u32 len);
void FixPkxChecksum(u32* pkx, u32 len);
inline u32 GetShuffleType(u32 pid);
void UnshufflePkx(u32* pkx_in, u32* pkx_out);
inline void SetUncryptedDataEqual(u32* pkx_in, u32* pkx_out);

int xy_decrypt(char* in_pkx, int in_pkx_length, char* out_pkx)
{
    u32 box_pkx[SIZE_BOX_PKX / sizeof(u32)];
    u32 outbox_pkx[SIZE_BOX_PKX / sizeof(u32)];
    u32 party_data[SIZE_PKX_PARTY_DATA / sizeof(u32)];
    
    Rnd xypkx;

    InitRnd(&xypkx);

    memcpy(box_pkx, in_pkx, in_pkx_length);

    if ((in_pkx_length != SIZE_PARTY_PKX) && (in_pkx_length != SIZE_BOX_PKX))
    {
        printf("Pkx is the wrong size.");
        return 0;
    }

    u32 decryptKey = box_pkx[0];    // key is now pid, not checksum

    CryptPkx(&box_pkx[2], &decryptKey, SIZE_PKX_CRYPT_LEN, &xypkx);
    UnshufflePkx(box_pkx, outbox_pkx);
    SetUncryptedDataEqual(box_pkx, outbox_pkx);

    if (!VerifyPkxChecksum(&box_pkx[2], (box_pkx[1] >> 16) & 0xFFFF, SIZE_PKX_CRYPT_LEN))
    {
        printf("Checksum failed.");
        return 0;
        //printf("Checksum failed.  Fixing.\n\n");
        //FixPkxChecksum(box_pkx, SIZE_PKX_CRYPT_LEN);
    }

    if (in_pkx_length == SIZE_PARTY_PKX)
    {

        decryptKey = box_pkx[0];

	memcpy(in_pkx + SIZE_BOX_PKX, party_data, SIZE_PKX_PARTY_DATA);

        CryptPkx(party_data, &decryptKey, SIZE_PKX_PARTY_DATA, &xypkx);
    }

    memcpy(out_pkx, outbox_pkx, SIZE_BOX_PKX);

    if (in_pkx_length == SIZE_PARTY_PKX)
    {
	memcpy(out_pkx + SIZE_BOX_PKX, party_data, SIZE_PKX_PARTY_DATA);
    }

    return in_pkx_length;
}


void CryptPkx(u32* pkx, u32* key, u32 len, Rnd* cr)
{
    u16* epkx = (u16*)pkx;
    u16 cryptKey = 0;
    
    for (u32 i = 0; i < len / 2; ++i)
    {
        cryptKey = CryptRNG(key, cr);
	printf("value %d, %d, %d, %d\n", epkx[i], cryptKey, epkx[i] ^ cryptKey, *key);
        epkx[i] ^= cryptKey;
    }
}


inline void InitRnd(Rnd* r)
{
    r->mul = 0x41C64E6D;
    r->add = 0x6073;
    r->rmul = 0xEEB9EB65;
    r->radd = 0xA3561A1;
}


inline u16 CryptRNG(u32* data, Rnd* r)
{
   *data = ((*data * r->mul) + r->add);
   return (*data >> 16);
}


inline u16 CalcPkxChecksum(u16* pkx, u32 len)
{
    u16 sum = 0;

    for (u32 i = 0; i < len / 2; ++i)
    {
        sum += pkx[i];

    }
    return sum;
}


bool VerifyPkxChecksum(u32* pkx, u16 currentSum, u32 len)       // this accepts the pkx starting @ the blocks
{
    u16* epkx = (u16*)pkx;

    u16 testSum = CalcPkxChecksum(epkx, len);

    printf("checksum %d, %d\n", currentSum, testSum);
    if (testSum == currentSum)
    {
        return true;
    }
    else
    {
        return false;
    }
}


void FixPkxChecksum(u32* pkx, u32 len)      // this accepts the whole pkx because it needs access to data before the blocks start
{
    u16* epkx = (u16*)pkx;

    u16 newSum = CalcPkxChecksum(&epkx[4], len);

    epkx[3] = newSum;
}


inline u32 GetShuffleType(u32 pid)      // same as it was before
{
    return ((pid & 0x3E000) >> 0xD) % 24;
}


void UnshufflePkx(u32* pkx_in, u32* pkx_out)
{
    u8* eipkx = (u8*)pkx_in;
    u8* eopkx = (u8*)pkx_out;

    u32 blockOrder = GetShuffleType(pkx_in[0]);

    for (u32 i = 0; i < 4; ++i)
    {
        memcpy(eopkx + SIZE_PKX_UNCRYPT + (SIZE_PKX_BLK * i), eipkx + SIZE_PKX_UNCRYPT + (SIZE_PKX_BLK * orders[blockOrder][i]), SIZE_PKX_BLK);
    }
}

inline void SetUncryptedDataEqual(u32* pkx_in, u32* pkx_out)
{
    for (u32 i = 0; i < 2; ++i)
    {
        pkx_out[i] = pkx_in[i];
    }
}


/*
pkx format:

u32 pid
u16 sanity
u16 checksum
56*4b pokedata blocks, shuffled same as always
28b party data 

*/
